mafia.py

```diff
@@ -1,110 +1,171 @@
     if len(username) > 0:
         return username
-    return None
 
 @app.route('/')
 @auth.login_required
 def index():
-    global player_id, ip2player
     username = str(auth.current_user())
-    role = ""
-    image_name = ""
     ip = str(request.remote_addr)
 
-    if ip in ip2player.keys():
-        return render_template("Player.html", player=ip2player[ip])
-    else:
-        if player_id > nPlayers:
-            return render_template("404.html", is_farsi=True)
-        role = roles[player_id]
-        image_name = role + "_" + str(randrange(1, nRoles[role] + 1))
-        ip2player[ip] = Player(ip, username, role, image_name)
-        print("*" * 20, "New Player","*" * 20)
-        toGod = ip + " : " + str(player_id) + " : " + username +  " --> " + role
-        toGod += "/" + role2fa[role]    #TODO: Just in Farsi Mode
-        print(toGod)
-        player_id += 1
+    # redirect current players
+    if player_manager.is_ip_valid(ip):
+        return render_template("Player.html", player=player_manager.get_player(ip))
+
+    # create new player
+    role = player_manager.get_next_role()
+    player : Player = player_manager.new_player(ip, username)
+    
+    # reject excess players
+    if player is None:
+        return not_found_page()
+
     return render_template("index.html",
-                            image_name=image_name,
-                            role_name=role, role_name_fa=role2fa[role],
-                            description=descriptions[role], description_fa=descriptions_fa[role],
+                            image_name=player.image_name,
+                            role_name=role.name, role_name_fa=role.get_name(is_farsi= True),
+                            description=role.get_description(), description_fa=role.get_description(is_farsi= True),
                             is_farsi=True)
 
-
 @auth_GOD.verify_password
 def verify_password_god(username, password):
     if password == preshared_key:
         return username
 
-
 @app.route('/GOD')
 @auth_GOD.login_required
-def GOD_PAGE():
-    global ip2player, nComments, comments_ordered
-    msg = ""
-    if request.args.get("Kill") is not None:
-        ip = request.args.get("Kill")
-        if ip in ip2player.keys():
-            if ip2player[ip].get_state() == "alive":
-                ip2player[ip].set_state("dead")
-            else:
-                ip2player[ip].set_state("alive")
-        else:
-            return render_template("404.html", is_farsi=True)
-    if request.args.get("Ban") is not None:
-        ip = request.args.get("Ban")
-        if ip in ip2player.keys():
-            if ip2player[ip].get_state() == "alive":
-                ip2player[ip].set_state("banned")
-            elif ip2player[ip].get_state() == "banned":
-                ip2player[ip].set_state("alive")
+def god_page():
+    def kill(player : Player) -> None:
+        # notice banned players cannot be killed
+        if player.is_alive():
+            player.die()
```

Comment:
add another `if` which makes player `alive` if `kill` is pressed again after player's death.

Reply:
done.
also checks for is_banned now